File: app_codebase/python_base_05/bin/core/ws/service/channel_registry.py
"""In-memory WebSocket channel registry (mirrors HTTP route registry)."""

from __future__ import annotations

from core.ws.contracts.register_channel_contract import ApplicationChannelSink, WsChannelHandler

_TIER_PUBLIC = "public"
_TIER_AUTHUSER = "authuser"
_TIER_SERVICE = "service"


class _ChannelRegistry(ApplicationChannelSink):
    def __init__(self) -> None:
        self._handlers: dict[str, WsChannelHandler] = {}

    def clear(self) -> None:
        self._handlers.clear()

    def _key(self, tier: str, channel: str) -> str:
        return f"{tier}:{channel}"

    def _add(self, tier: str, channel: str, handler: WsChannelHandler) -> None:
        self._handlers[self._key(tier, channel)] = handler

    def get_handler(self, tier: str, channel: str) -> WsChannelHandler | None:
        return self._handlers.get(self._key(tier, channel))

    def public_channel(self, channel: str, handler: WsChannelHandler) -> None:
        self._add(_TIER_PUBLIC, channel, handler)

    def authuser_channel(self, channel: str, handler: WsChannelHandler) -> None:
        self._add(_TIER_AUTHUSER, channel, handler)

    def service_channel(self, channel: str, handler: WsChannelHandler) -> None:
        self._add(_TIER_SERVICE, channel, handler)


_the_registry = _ChannelRegistry()
application_ws_sink: ApplicationChannelSink = _the_registry


def reset_channel_registry() -> None:
    _the_registry.clear()


def get_channel_handler(tier: str, channel: str) -> WsChannelHandler | None:
    return _the_registry.get_handler(tier, channel)

File: app_codebase/dart_bkend_base_02/bin/core/ws/service/channel_registry.dart
/// In-memory WebSocket channel registry (mirrors HTTP route registry).
library;

import '../contracts/register_channel_contract.dart';
import '../contracts/ws_message_contract.dart';

const _tierPublic = 'public';
const _tierAuthuser = 'authuser';
const _tierService = 'service';

final ApplicationChannelSink applicationWsSink = _ChannelRegistry._instance;

void resetChannelRegistry() => _ChannelRegistry._instance.clear();

WsChannelHandler? getChannelHandler(String tier, String channel) =>
    _ChannelRegistry._instance.getHandler(tier, channel);

class _ChannelRegistry implements ApplicationChannelSink {
  _ChannelRegistry._();

  static final _ChannelRegistry _instance = _ChannelRegistry._();

  final Map<String, WsChannelHandler> _handlers = {};

  void clear() => _handlers.clear();

  String _key(String tier, String channel) => '$tier:$channel';

  void _add(String tier, String channel, WsChannelHandler handler) {
    _handlers[_key(tier, channel)] = handler;
  }

  WsChannelHandler? getHandler(String tier, String channel) =>
      _handlers[_key(tier, channel)];

  @override
  void publicChannel(String channel, WsChannelHandler handler) =>
      _add(_tierPublic, channel, handler);

  @override
  void authuserChannel(String channel, WsChannelHandler handler) =>
      _add(_tierAuthuser, channel, handler);

  @override
  void serviceChannel(String channel, WsChannelHandler handler) =>
      _add(_tierService, channel, handler);
}
